src/gtk/overlay.cpp

```diff
@@ -0,0 +1,222 @@
+/////////////////////////////////////////////////////////////////////////////
+
+// ============================================================================
+// declarations
+// ============================================================================
+
+// ----------------------------------------------------------------------------
+// headers
+// ----------------------------------------------------------------------------
+
+// For compilers that support precompilation, includes "wx.h".
+#include "wx/wxprec.h"
+
+
+#ifndef WX_PRECOMP
+    #include "wx/log.h"
+    #include "wx/window.h"
+    #include "wx/dcclient.h"
+#endif
+
+#include "wx/private/overlay.h"
+#include "wx/gtk/dc.h"
+#include "wx/gtk/private.h"
+
+#ifdef wxHAS_NATIVE_OVERLAY
+
+#define wxDEBUG_GTK_OVERLAY 0 // set this to 1 to see the overlay
+
+// ============================================================================
+// implementation
+// ============================================================================
+extern "C" {
+static gboolean
+wxgtk_draw_overlay(GtkWidget* widget, cairo_t* cr, wxOverlayImpl* overlay)
+{
+    if ( gtk_cairo_should_draw_window(cr, gtk_widget_get_window(widget)) )
+    {
+        cairo_surface_t* surface = overlay->GetSurface();
+
+        if ( surface && cairo_surface_status(surface) == CAIRO_STATUS_SUCCESS )
+        {
+            cairo_set_source_surface(cr, surface, 0, 0);
+            cairo_paint(cr);
+        }
+    }
+
+    return false;
+}
+}
+
+namespace // anonymous
+{
+inline GtkWidget* GetOverlayWidget(wxWindow* win)
+{
+    if ( !win )
+        return NULL;
+
+    // The overlay widget (if enabled for this win) is either m_widget
+    // or m_wxwindow's parent.
+
+    GtkWidget* overlay = win->m_widget;
+
+    if ( GTK_IS_OVERLAY(overlay) )
+        return overlay;
+
+    overlay = gtk_widget_get_parent(win->m_wxwindow);
+
+    if ( GTK_IS_OVERLAY(overlay) )
+        return overlay;
+
+    return NULL;
+}
+} // anonymous namespace
+// ----------------------------------------------------------------------------
+// wxOverlay
+// ----------------------------------------------------------------------------
+
+wxOverlayImpl::wxOverlayImpl()
+{
+    m_surface = NULL;
+    m_window = NULL;
+    m_cairoContext = NULL;
+    m_cairoSurface = NULL;
+}
+
+wxOverlayImpl::~wxOverlayImpl()
+{
+    if ( m_window )
+        m_window->Unbind(wxEVT_SIZE, &wxOverlayImpl::OnSize, this);
+
+    if ( m_cairoSurface )
+        cairo_surface_destroy(m_cairoSurface);
+
+    if ( m_surface )
+    {
+        gtk_widget_destroy(m_surface);
+        g_object_unref(m_surface);
+    }
+}
+
+bool wxOverlayImpl::IsOk() const
+{
+    return m_window != NULL;
+}
+
+void wxOverlayImpl::Init(wxDC *dc, int x, int y, int width, int height)
+{
+    wxCHECK_RET( dc, "NULL dc pointer" );
+    wxASSERT_MSG( !IsOk() , "You cannot Init an overlay twice" );
+
+    wxLogDebug("wxOverlayImpl::Init(%d, %d, %d, %d)", x, y, width, height);
```

Comment:
Please remove this in the final version or replace it with `wxLogTrace("overlay", ...)` if you think it can be useful to keep it.